File: mlp-ga/scripts/Mlp.js
class Mlp {

    constructor(numberOfInputs, neuronsPerLayer, neuronLayers) {
        if(neuronLayers === null) {
            this.neuronLayers = [];
            for (var i = 0; i < neuronsPerLayer.length; i++) {
                var neuronLayer = this.initializeNeuronLayer(numberOfInputs, neuronsPerLayer[i], i);
                this.neuronLayers.push(neuronLayer);
            }
        } else {
            this.neuronLayers = neuronLayers;
        }
    }

    initializeNeuronLayer(numberOfInputs, numberOfNeurons, numberOflayer) {
        var ret = [];
        for(var i = 0; i < numberOfNeurons; i++) {
            var neuron;
            if(numberOflayer === 0) {
                neuron = new Neuron(numberOfInputs, null, null);
            } else {
                var middleLayerNumberOfInputs = this.neuronLayers[numberOflayer - 1].length;
                neuron = new Neuron(middleLayerNumberOfInputs, null, null);
            }
            ret.push(neuron);
        }
        return ret;
    }

    calculateOutputs(inputs) {
        for(var i = 0; i < this.neuronLayers.length; i++) {
            for(var j = 0; j < this.neuronLayers[i].length; j++) {
                if(i === 0) {
                    this.neuronLayers[i][j].calculateOutput(inputs);
                } else {
                    var middleLayerInputs = this.getLayerOutputs(this.neuronLayers[i - 1]);
                    this.neuronLayers[i][j].calculateOutput(middleLayerInputs);
                }
            }
        }
        return this.getLayerOutputs(this.neuronLayers[this.neuronLayers.length - 1]);
    }

    getLayerOutputs(neuronLayer) {
        var ret = [];
        for(var i = 0; i < neuronLayer.length; i++) {
            ret.push(neuronLayer[i].getOutput());
        }
        return ret;
    }

    print() {
        for(var i = 0; i < this.neuronLayers.length; i++) {
            for(var j = 0; j < this.neuronLayers[i].length; j++) {
                console.log("[" + i + "][" + j + "]");
                this.neuronLayers[i][j].print();
            }
        }
    }

    static reproduce(mlp1, mlp2, mutationRate) {
        var neuronLayers = [];
        for(var i = 0; i < mlp1.neuronLayers.length; i++) {
            var neuronLayer = [];
            for(var j = 0; j < mlp1.neuronLayers[i].length; j++) {
                var neuron = Neuron.reproduce(mlp1.neuronLayers[i][j], mlp2.neuronLayers[i][j], mutationRate);
                neuronLayer.push(neuron)
            }
            neuronLayers.push(neuronLayer);
        }
        return new Mlp(null, null, neuronLayers);
    }

}

File: libraries/mlp/src/main/java/com/yzumis/mlp/Mlp.java
package com.yzumis.mlp;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Yzumi on 27/12/2017.
 */
public class Mlp {

    private final List<List<Neuron>> neuronLayers;

    public Mlp(final int numberOfInputs, final List<Integer> neuronsPerLayer) {
        this.neuronLayers = new ArrayList<>();
        for (int i = 0; i < neuronsPerLayer.size(); i++) {
            final List<Neuron> layer = this.initializeNeuronLayer(numberOfInputs, neuronsPerLayer.get(i), i);
            this.neuronLayers.add(layer);
        }
    }

    public Mlp(final List<List<Neuron>> neuronLayers) {
        this.neuronLayers = neuronLayers;
    }

    private List<Neuron> initializeNeuronLayer(final int numberOfInputs, final int neuronsInLayer, final int numberOflayer) {
        final List<Neuron> ret = new ArrayList<>();
        for(int i = 0; i < neuronsInLayer; i++) {
            final Neuron neuron;
            if(numberOflayer == 0) {
                neuron = new Neuron(numberOfInputs);
            } else {
                final int middleLayerNumberOfInputs = this.neuronLayers.get(numberOflayer - 1).size();
                neuron = new Neuron(middleLayerNumberOfInputs);
            }
            ret.add(neuron);
        }
        return ret;
    }

    public List<Float> calculateOutputs(final List<Float> inputs) {
        for(int i = 0; i < this.neuronLayers.size(); i++) {
            for(int j = 0; j < this.neuronLayers.get(i).size(); j++) {
                if( i == 0) {
                    this.neuronLayers.get(i).get(j).calculateOutput(inputs);
                } else {
                    final List<Float> middleLayerInputs = this.getLayerOutputs(this.neuronLayers.get(i - 1));
                    this.neuronLayers.get(i).get(j).calculateOutput(middleLayerInputs);
                }
            }
        }
        return this.getLayerOutputs(this.neuronLayers.get(this.neuronLayers.size() -1));
    }

    private List<Float> getLayerOutputs(final List<Neuron> neuronLayer) {
        final List<Float> ret = new ArrayList<>();
        for(final Neuron neuron : neuronLayer) {
            ret.add(neuron.getOutput());
        }
        return ret;
    }

    public static Mlp reproduce(final Mlp mlp1, final Mlp mlp2, final float mutationRate) {
        final List<List<Neuron>> childNeuronLayers = new ArrayList<>();
        for(int i = 0; i < mlp1.neuronLayers.size(); i++) {
            final List<Neuron> childNeuronLayer = new ArrayList<>();
            for(int j = 0; j < mlp1.neuronLayers.get(i).size(); j++) {
                final Neuron neuron = Neuron.reproduce(mlp1.neuronLayers.get(i).get(j), mlp2.neuronLayers.get(i).get(j), mutationRate);
                childNeuronLayer.add(neuron);
            }
            childNeuronLayers.add(childNeuronLayer);
        }
        return new Mlp(childNeuronLayers);
    }
}
